### PR TITLE
Fix UUID links in writer field in subfolder setups

### LINKED ISSUE #5208: [v4] Writer field: UUID links are removed in subfolder installations
## Description

Add a link to a page in writer using the link field, adds `<a>Text</a>` without href.

## Notes for later reference:

Happens because of "The URL points outside of the site index URL" check in `Dom::isAllowedUrl()`.

Possible solution: We do have the `allowedDomains` option already, but that of course only covers domains. We could make `allowHostRelativeUrls` with a default of `true`, but then if someone needs the protection, the writer field would again not work.

Discord discussion: https://discord.com/channels/525634039965679616/525637437494657024/1110583259919818935

### PR DESCRIPTION
**Pull request by lukasbestle**
### Enhancements

- Dom: New `allowHostRelativeUrls` sanitization option (`true` by default) to perform less strict checks when the HTML `<base>` element is used
- Dom: Custom sanitization callbacks for attributes, elements and the doctype now also receive the `$options` array
- Sane: New `$isExternal` mode specifically for external files that may be accessed directly

### Fixes

- UUID permalinks added via the link dialog of the writer field are no longer removed in subfolder setups #5208

### Breaking changes

- When sanitizing DOM objects (e.g. in the writer field, but not during the sanitization/validation of uploaded files), host-relative URLs that point outside the site root are now allowed as the use of the HTML `<base>` element is assumed for sites in a subfolder. To revert to the old, strict behavior, set the `allowHostRelativeUrls` option to `false`.
- New `$isExternal` argument for the `Sane\Handler::sanitize()` and `::validate()` methods that custom Sane handlers need to implement; it allows to differentiate between strings from external files that may be accessed directly and strings that will end up directly on the page

## Ready?

- [x] Unit tests for fixed bug/feature
- [x] In-code documentation (wherever needed)
- [x] Tests and checks all pass

### For review team

- [ ] Add changes to release notes draft in Notion
- [x] ~Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)~

### CODE REVIEW

File: src/Sane/DomHandler.php [hunk 1]
```diff
@@ -43,6 +43,15 @@
...
 	public static array|bool $allowedDomains = true;
 
+	/**
+	 * Whether URLs that begin with `/` should be allowed even if the
+	 * site index URL is in a subfolder (useful when using the HTML
+	 * `<base>` element where the sanitized code will be rendered)
+	 *
+	 * @SuppressWarnings(PHPMD.LongVariable)
+	 */
+	public static bool $allowHostRelativeUrls = true;
```
[the remark this answers is not in the document]
Comment (lukasbestle) [reply src/Sane/DomHandler.php:53]: Hm, looks like the suppress syntax is not supported for prop definitions. I'll look into this again tomorrow.

[the remark this answers is not in the document]
Comment (lukasbestle) [reply src/Sane/DomHandler.php:53]: Do you have an idea for a shorter option name? Or is it OK to suppress this warning?

**Comment by lukasbestle** [comment]
PR is ready for review, but needs more unit tests for the Sane changes before it can be merged. I will add the missing tests this evening.

File: src/Sane/DomHandler.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (bastianallgeier) [reply src/Sane/DomHandler.php:53]: allowRelativeHosts maybe? But it's not that much shorter either.

[the remark this answers is not in the document]
Comment (bastianallgeier) [reply src/Sane/DomHandler.php:53]: Thought about it again and your name is definitely better. Let's ignore the warning.

**Review by bastianallgeier — APPROVED** [review]

**Pull request merged**